### ISSUE TITLE
Compilation Error when added dependency of FsLab

### ISSUE THREAD

**Issue by garrardkitchen** [Azure/azure-webjobs-sdk-script#650]
Hi,

I'm attempting to use FsLab (http://fslab.org/) to create and return HTML chart via Azure Function however I'm getting an error.
#### Repro steps

Provide the steps required to reproduce the problem
1. Create project.json with

```
{
  "frameworks": {
    "net46":{
      "dependencies": {
        "FSharp.Data": "2.3.2",
        "Newtonsoft.Json": "9.0.1",
        "FsLab" : "1.0.0",
        "RProvider" : "1.1.20"
      }
    }
   }
}
```
1. Wait for compilation
#### Expected behavior

Not error
#### Actual behavior

Logs:

```
2016-09-07T10:40:49.424 Restoring packages.
2016-09-07T10:40:49.424 Starting NuGet restore
2016-09-07T10:40:51.004 Restoring packages for D:\home\site\wwwroot\api-chart\project.json...
2016-09-07T10:40:52.837 GET https://api.nuget.org/v3-flatcontainer/http.fs/index.json
2016-09-07T10:40:54.184 
2016-09-07T10:40:54.465 Packages restored.
2016-09-07T10:40:55.004 Script for function 'api-chart' changed. Reloading.
2016-09-07T10:40:56.010 OK https://api.nuget.org/v3-flatcontainer/http.fs/index.json 459ms
2016-09-07T10:40:56.010 GET https://api.nuget.org/v3-flatcontainer/http.fs/0.8.0/http.fs.0.8.0.nupkg
2016-09-07T10:40:56.010 OK https://api.nuget.org/v3-flatcontainer/http.fs/0.8.0/http.fs.0.8.0.nupkg 399ms
2016-09-07T10:40:56.010 WARNING: Detected package downgrade: Newtonsoft.Json from 0.0.0 to 9.0.1
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> FsLab (>= 1.0.0) -> XPlot.GoogleCharts (= 1.3.1) -> Newtonsoft.Json
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> Newtonsoft.Json (>= 9.0.1)
2016-09-07T10:40:56.010 WARNING: Detected package downgrade: Google.DataTable.Net.Wrapper from 0.0.0 to 3.1.2.0
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> FsLab (>= 1.0.0) -> XPlot.GoogleCharts (= 1.3.1) -> Google.DataTable.Net.Wrapper
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> FsLab (>= 1.0.0) -> Google.DataTable.Net.Wrapper (= 3.1.2)
2016-09-07T10:40:56.010 WARNING: Detected package downgrade: Deedle from 0.0.0 to 1.2.5
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> FsLab (>= 1.0.0) -> XPlot.GoogleCharts.Deedle (= 1.3.1) -> Deedle
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> FsLab (>= 1.0.0) -> Deedle (= 1.2.5)
2016-09-07T10:40:56.010 WARNING: Detected package downgrade: Newtonsoft.Json from 0.0.0 to 9.0.1
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> FsLab (>= 1.0.0) -> XPlot.Plotly (= 1.3.1) -> Newtonsoft.Json
2016-09-07T10:40:56.010 api-chart (>= 1.0.0) -> Newtonsoft.Json (>= 9.0.1)
2016-09-07T10:40:56.010 Committing restore...
2016-09-07T10:40:56.010 Writing lock file to disk. Path: D:\home\site\wwwroot\api-chart\project.lock.json
2016-09-07T10:40:56.010 D:\home\site\wwwroot\api-chart\project.json
2016-09-07T10:40:56.010 Restore completed in 3206ms.
2016-09-07T10:40:56.010 
2016-09-07T10:40:56.010 NuGet Config files used:
2016-09-07T10:40:56.010 C:\DWASFiles\Sites\speech-svrless\AppData\NuGet\NuGet.Config
2016-09-07T10:40:56.010 
2016-09-07T10:40:56.010 Feeds used:
2016-09-07T10:40:56.010 https://api.nuget.org/v3/index.json
2016-09-07T10:40:56.010 
2016-09-07T10:41:00.603 startup(2,1): error FS3053: The type provider 'RProvider.RProvider' reported an error: The type provider constructor has thrown an exception: Could not load file or assembly 'RDotNet, Version=1.6.5.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The system cannot find the file specified.
2016-09-07T10:41:00.603 startup(2,1): warning FS3005: Referenced assembly 'D:\home\data\Functions\packages\nuget\RProvider\1.1.20\lib\net40\RProvider.dll' has assembly level attribute 'Microsoft.FSharp.Core.CompilerServices.TypeProviderAssemblyAttribute' but no public type provider classes were found
2016-09-07T10:41:00.603 startup(2,1): error FS3053: The type provider 'RProvider.RProvider' reported an error: The type provider constructor has thrown an exception: Could not load file or assembly 'RDotNet, Version=1.6.5.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The system cannot find the file specified.
2016-09-07T10:41:00.603 startup(2,1): warning FS3005: Referenced assembly 'D:\home\data\Functions\packages\nuget\RProvider\1.1.20\lib\net40\RProvider.dll' has assembly level attribute 'Microsoft.FSharp.Core.CompilerServices.TypeProviderAssemblyAttribute' but no public type provider classes were found
2016-09-07T10:41:00.603 Compilation failed.
```

Warmest regards,

Garrard.

**Comment by fabiocav** [comment]
@tpetricek do the type providers expect the assemblies to be loaded in a specific way?

**Comment by tpetricek** [comment]
Yes, they do - they get loaded and called by the F# compiler. The compiler loads them from the location where the package manager puts them. The R type provider specifically, as part of the loading, registers `AssemblyResolve` handler and in the handler looks for R.NET (another package, also from NuGet) in "typical folders used by NuGet".

By default, it searches the path specified in `ProbingLocations` in `app.config` and we ship default config file that works for NuGet and Paket: https://github.com/BlueMountainCapital/FSharpRProvider/blob/master/src/RProvider/app.config#L4

It seems that we need to add something to the probing locations to make it use with whatever Azure functions are using. Assume I install RProvider and RDotNet from NuGet, what will be the relative path to `RDotNet.dll` from `RProvider.dll`? In Paket and NuGet, this is `../../../*/lib/net40`. 

For Azure Functions - based on the sample above, I would guess we need one more `../` because the version number is adding one more nested folder?

tl;dr @garrardkitchen This is a bit of a mess but we'll sort it out :) In the meantime, can you reference just the libraries you need? (Which, I suppose is FSharp.Data and XPlot?)

**Comment by tpetricek** [comment]
@fabiocav I'll be able to look into this next week - if you want more background, we actually have very detailed [documentation on how this works](http://bluemountaincapital.github.io/FSharpRProvider/devnotes.html) and [instructions for debugging R provider issues](http://bluemountaincapital.github.io/FSharpRProvider/diagnostics.html) (As you can see, this is mostly black magic :-))

**Comment by garrardkitchen** [comment]
Hi @tpetricek, I will definitely try this and let you know how I get on.  Thank you for looking into this BTW. It may now be next week though before I get to this.

**Comment by lindydonna** [comment]
@tpetricek I'm moving this into Next - Triaged. CC @dsyme and @sylvanc

**Comment by garrardkitchen** [comment]
Hi @tpetricek, I changed my sample code to use XPlot and FSharp.Data and it compiles but I get an error when I access the uri.

error:

```
2016-11-03T15:41:16.428 Compilation succeeded.
2016-11-03T15:41:45.747 Function started (Id=dbad1aab-d34d-4099-8c1f-eb861d35ca14)
2016-11-03T15:41:45.747 Generate chart
2016-11-03T15:41:45.762 Function completed (Failure, Id=dbad1aab-d34d-4099-8c1f-eb861d35ca14)
2016-11-03T15:41:45.996 Exception while executing function: Functions.api-chart. XPlot.GoogleCharts: Could not load file or assembly 'Google.DataTable.Net.Wrapper, Version=3.1.2.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The system cannot find the file specified.
2016-11-03T15:43:15  No new trace in the past 1 min(s).
```

run.fsx:
```
#if !COMPILED
#I @"../../bin/Binaries/WebJobs.Script.Host"
#r "Microsoft.Azure.WebJobs.Host.dll"
#r "System.Web.Http.dll"
#r "System.Net.Http.dll"
#r "System.IO.dll"
#endif

#r "System.Data.dll"
#r "Google.DataTable.Net.Wrapper.dll"
#r "XPlot.GoogleCharts.dll"
#r "XPlot.Plotly.dll"

open System
open FSharp.Data
open Google.DataTable.Net.Wrapper
open XPlot.GoogleCharts
open XPlot.Plotly
open System.Net
open System.Net.Http
open Microsoft.Azure.WebJobs.Host
open System.Data
open System.Threading.Tasks

type data = {
    label: string
    value: float
}

let Run (req: HttpRequestMessage, log: TraceWriter) : Task<HttpResponseMessage> =  

    async { 
        log.Info(sprintf "Generate chart")

        let mutable Bolivia = [("2015/2016",2);("2016/2017",2);("2017/2018",19)]
        let mutable Ecuador = [("2015/2016",6);("2016/2017",42);("2017/2018",3)]
        let mutable Madagascar = [("2015/2016",11);("2016/2017",22);("2017/2018",2)]
        let mutable Average = [("2015/2016",31);("2016/2017",3);("2017/2018",14)]


        let series = [ "bars"; "bars"; "bars"; "lines" ]
        let inputs = [ Bolivia; Ecuador; Madagascar; Average ]

        let output =
            inputs
            |> Chart.Combo
            |> Chart.WithOptions 
                (Options(title = "Coffee Production", series = 
                    [| for typ in series -> Series(typ) |]))
            |> Chart.WithLabels 
                ["Bolivia"; "Ecuador"; "Madagascar"; "Average"]
            |> Chart.WithLegend true
            |> Chart.WithSize (600, 250)   

        let res = HttpStatusCode.OK |> HttpResponseMessage
        res.Content <- new StringContent(output.Html)
        return res  
    } |> Async.StartAsTask
```

project.json;
```
{
  "frameworks": {
    "net46":{
      "dependencies": {
        "FSharp.Data": "2.3.2",
        "XPlot.Plotly" : "1.3.1",        
        "XPlot.GoogleCharts" : "1.3.1",
        "Google.DataTable.Net.Wrapper" : "3.1.2"      
      }
    }
   }
}
```

Warmest regards,

Garrard.

**Comment by dsyme** [comment]
@garrardkitchen @lindydonna @fabiocav @tpetricek @christopheranderson - We are actively looking into this bug this week.  We think nearly all the existing F# issues boil down to the one same bug (not finding assemblies at runtime which have been resolved statically)

**Comment by fabiocav** [comment]
Great! Let me know if you guys need to chat or  support with anything.

**Comment by dsyme** [comment]
@fabiocav I've pinged you with a few questions about CI.  If we could have permission to add the full-ci tag ourselves that would be really useful, as submitting failing CI tests are the best way for us to be sure we're nailing this.

**Comment by sylvanc** [comment]
This has been partially fixed, in that FsLab is no longer problematic, and RProvider can find RDotNet.

RProvider still does not load, of course, since R is not installed. This is expected behaviour. The error message from RProvider isn't ideal, as it complains about a broken pipe rather than R not being installed.

Depending on FsLab now works.

**Comment by dsyme** [comment]
@fabiocav @lindydonna This can be closed, per @sylvanc's comment above, thanks

**Comment by mathewc** [comment]
Closed, thanks.
